src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +141,242 @@ public interface IFileWatchingToken
+                        return true;
+
+                    case Kind.UnwatchDirectories when other._kind == Kind.UnwatchDirectories:
+                        var cookies = new List<uint>(_cookies);
+                        cookies.AddRange(other._cookies);
+                        combined = UnwatchDirectories(cookies);
+                        return true;
+
+                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFile:
+                        combined = UnwatchFiles(_tokens.ToImmutableList().Add(other._token));
+                        return true;
+
+                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFiles:
+                        combined = UnwatchFiles(_tokens.ToImmutableList().AddRange(other._tokens));
+                        return true;
+
+                    default:
+                        break;
+                }
+
+                combined = default;
+                return false;
+            }
+
+            public async ValueTask ApplyAsync(IVsAsyncFileChangeEx service, CancellationToken cancellationToken)
+            {
+                switch (_kind)
+                {
+                    case Kind.None:
+                        return;
+
+                    case Kind.WatchDirectory:
+                        var cookie = await service.AdviseDirChangeAsync(_directory, watchSubdirectories: true, _sink, cancellationToken).ConfigureAwait(false);
+                        _cookies.Add(cookie);
+
+                        if (_filter != null)
+                            await service.FilterDirectoryChangesAsync(cookie, new string[] { _filter }, cancellationToken).ConfigureAwait(false);
+
+                        return;
+
+                    case Kind.WatchFile:
+                        ((Context.FileWatchingToken)_token).Cookie = await service.AdviseFileChangeAsync(_directory, _fileChangeFlags, _sink, cancellationToken).ConfigureAwait(false);
+                        return;
+
+                    case Kind.UnwatchFile:
+                        await service.UnadviseFileChangeAsync(((Context.FileWatchingToken)_token).Cookie!.Value, cancellationToken).ConfigureAwait(false);
+                        return;
+
+                    case Kind.UnwatchDirectories:
+                        Contract.ThrowIfFalse(_cookies is not null);
+                        foreach (var unwatchCookie in _cookies)
+                        {
+                            await service.UnadviseDirChangeAsync(unwatchCookie, cancellationToken).ConfigureAwait(false);
+                        }
+
+                        return;
+
+                    case Kind.UnwatchFiles:
+                        Contract.ThrowIfFalse(_tokens is not null);
+                        foreach (var token in _tokens)
+                        {
+                            await service.UnadviseFileChangeAsync(((Context.FileWatchingToken)token).Cookie!.Value, cancellationToken).ConfigureAwait(false);
+                        }
```

Comment:
A new API was just added -- IVsAsyncFileChangeEx.UnadviseFileChangesAsync -- that can unadvise all of the tokens' cookies at once.  If that's available to you, it would be better to use it here.  If not, consider adding a comment here to use the batch operation when it becomes available.

Ditto for the UnwatchDirectories case (using UnadviseDirChangesAsync).

Reply:
I'm going to submit this part as a separate change to a branch where the newer API is available. This pull request restructures the feature to minimize the code change between the two development branches.